butane_core/src/db/mod.rs

```
@@ -560,28 +560,116 @@ impl ConnectionSpec {
             None => Err(crate::Error::UnknownBackend(self.backend_name.clone())),
         }
     }
+    /// Get the backend name.
+    pub const fn backend_name(&self) -> &String {
+        &self.backend_name
+    }
+    /// Get the database connection string.
+    pub const fn connection_string(&self) -> &String {
+        &self.conn_str
+    }
+    /// Get the connection string URI if it is a URI.
+    /// Returns None if the connection string is not a URI.
+    pub fn connection_string_uri(&self) -> Option<url::Url> {
+        url::Url::parse(self.connection_string()).ok()
+    }
+
+    fn is_pg_key_value_pairs(connection_string: &str) -> bool {
+        // host and hostaddr are optional in Postgresql, however this prevents connecting
+        // without one of them: https://github.com/sfackler/rust-postgres/issues/1239
+        connection_string.contains('=')
+            && (connection_string.contains("host=")
+                || connection_string.contains("hostaddr=")
+                || connection_string.contains("port=")
+                || connection_string.contains("dbname=")
+                || connection_string.contains("user=")
+                || connection_string.contains("host =")
+                || connection_string.contains("hostaddr =")
+                || connection_string.contains("port =")
+                || connection_string.contains("dbname =")
+                || connection_string.contains("user ="))
+    }
+
+    /// Add a query parameter to the connection string.
+    pub fn add_parameter(&mut self, name: &str, value: &str) {
+        if Self::is_pg_key_value_pairs(&self.conn_str) {
+            // Append using PostgreSQL key-value pair syntax.
+            self.conn_str.push_str(&format!(" {}={}", name, value));
+            return;
+        }
+        if self.conn_str.contains('?') {
```

I think we should check if the connection string is a URL first. If it's not a URL and it's not postgres key-value pairs then we don't really know how to add a parameter and should return an error

---

I'll check that, but IIRC SQLite supported `?` for things that were not a valid URI